### ISSUE TITLE
Open method creates new AssetsAudioPlayer

### ISSUE THREAD

**Issue by johnny-stevie** [florent37/Flutter-AssetsAudioPlayer#277]
**Flutter Version**

My version : 1.18.0-10.0.pre

**Lib Version**

My version : ^2.0.5

**Platform (Android / iOS / web) + version**

Platform : Android 10

I want to start off by saying, thank you for this amazing library, I've been looking for something like this for months now.

**Bug**
When I have a list of songs on ListTiles with an onTap callback to the .open() function. If the feature is misused by tapping on lots of them in a short amount of time, new Audio Players are spawned playing songs in parallel which is not what I want to happen.
Is there a way to stop new AudioPlayers from being created? Is it something with the open function not being given enough time to respond when tapped in a short period of time? 

**Things I've Tried but didn't help**
-- using Future.delay() before calling .open()
--monitoring the map of audio players with AssetsAudioPlayer.allPlayers() and disposing of new instances of AudioPlayers

Any help is appreciated!

**Small code to reproduce**

```dart
import 'package:assets_audio_player/assets_audio_player.dart';
import 'package:flutter/material.dart';

//sketch up code to show bug in a list 

void main() {
  runApp(MyApp());
}

class MyApp extends StatefulWidget {
  
  const MyApp({Key key, this.playlist}) : super(key: key);
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State {
  
  final AssetsAudioPlayer _assetsAudioPlayer = AssetsAudioPlayer();

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        body: List.builder(
          itemCount:widget.playlist.audios.length,
          itemBuilder:(BuildContext context, int index){
            return ListTile(
               title: Text("open"),
                onTap: () {
                  //open code here
               _assetsAudioPlayer.open(
                  widget.playlist,
               )
            }
          ),
        ),
      ),
    );
  }
}
```

**Comment by liuchuancong** [comment]
```
assetsAudioPlayer = AssetsAudioPlayer.withId('flutter_assets_audio_player');
```

**Comment by johnny-stevie** [comment]
I've already tried creating an audio player with a unique Id and deleting
new entries in the map, it doesn't work. It's because if you tap on a lot
of ListTiles at once, the .open() function gets overwhelmed and the error
handler creates new audio players. That's what I think happens.

**Comment by florent37** [comment]
open should not create a new player,

but as said @liuchuancong you can use .withId should fetch the same previous used player

but maybe using a lot of time the .open method (and maybe dispose ?) create/remove the player

I'll try your code, maybe I missed something

**Comment by johnny-stevie** [comment]
@florent37 Thanks for the quick reply. I've just tried using @liuchuancong 's suggestion and it didn't work. The songs kept on simultaneously getting played by new AudioPlayers. Just to let you know, the code I provided is the simplified  version, since my original list heavily depends on API calls.

**Comment by liuchuancong** [comment]
```
 Future<void> initPlaylist(Playlist playlist) async {
    if (isPlay) {
      stop();
    }
    try {
      await assetsAudioPlayer.open(playlist,
          showNotification: true, autoStart: false);
    } catch (t) {
      //mp3 unreachable
    }
  }
```

after done, you  should  use 


```
  Future<void> reMoveAtIndex(int index) async {
    print(assetsAudioPlayer.playlist.audios.length);
    assetsAudioPlayer.playlist.removeAtIndex(index);
  }

  Future<void> playlistPlayAtIndex(int index) async {
    await assetsAudioPlayer.playlistPlayAtIndex(index);
  }

  Future<void> seek(Duration by) async {
    await assetsAudioPlayer.seek(by);
  }

  Future<void> playOrPause() async {
    await assetsAudioPlayer.playOrPause();
  }

  Future<void> play() async {
    await assetsAudioPlayer.play();
  }

  Future<void> pause() async {
    await assetsAudioPlayer.pause();
  }

  Future<void> stop() async {
    await assetsAudioPlayer.stop();
  }

  Future<void> next() async {
    await assetsAudioPlayer.next();
  }

  Future<void> prev() async {
    await assetsAudioPlayer.previous();
  }
```

if  you want  play anthor Playlist, you can use initPlaylist method again.

**Comment by johnny-stevie** [comment]
@liuchuancong  This doesn't entirely stop the problem as new AudioPlayers are still spawned but they are stopped from playing. That still takes up more space than needed. I need there to be only 1 AudioPlayer at any one time. Also, with this way of doing it, the user has to manually go to the notification to start the audio, instead of the onTap function straight away dealing with it.

**Comment by pparadox11** [comment]
Hello, i believe @johnny-stevie is right, i have created a single instance of an AssetAudioPlayer,
then after in my code i call it like this,

`      
try {
        await assetsAudioPlayer.open(
          Audio('assets/claim_trip.mp3'),
          autoStart: false,
        );

//        await assetsAudioPlayer.play();
      } on PlatformException {}
`

and 

`
Future.delayed(
   const Duration(seconds: 10), () => assetsAudioPlayer.stop());
`

but the player doesn't stop. When i remove the commented line, then it starts 2 players, and the 1 stops after 10 seconds.

**Comment by johnny-stevie** [comment]
I've tried the delayed function, no luck there. I think it's to do with how .open handles multiple requests at the same time that's the problem.

**Comment by liuchuancong** [comment]
```
import 'package:assets_audio_player/assets_audio_player.dart';

class AudioInstance {
  factory AudioInstance() => _getInstance();
  bool get isPlay => assetsAudioPlayer.isPlaying.value;
  Playlist get playList => assetsAudioPlayer.playlist;
  static AudioInstance _instance;
  static AudioInstance get instance => _getInstance();
  AssetsAudioPlayer assetsAudioPlayer;
  AudioInstance._internal() {
    assetsAudioPlayer = AssetsAudioPlayer.withId('flutter_assets_audio_player');
  }

  static AudioInstance _getInstance() {
    if (_instance == null) {
      _instance = AudioInstance._internal();
    }
    return _instance;
  }

  static Audio _audio;


  Future<void> playFileAudio(String url, song) async {
    if (isPlay) {
      stop();
    }

    try {
      _audio = Audio.file(
        url,
        metas: Metas(
          title: song.name,
          artist: song.artists[0].name,
          album: song.album.name,
          image:
              MetasImage.network(song.album.picUrl), //can be MetasImage.network
        ),
      );
      await assetsAudioPlayer.open(_audio, showNotification: true);
      updateMetas(song);
    } catch (t) {
      //mp3 unreachable
    }
  }

  Future<void> initPlaylist(Playlist playlist) async {
    if (isPlay) {
      stop();
    }
    try {
      await assetsAudioPlayer.open(playlist,
          showNotification: true, autoStart: false);
    } catch (t) {
      //mp3 unreachable
    }
  }

  updateMetas(song) {
    if (_audio != null) {
      _audio.updateMetas(
        title: song.name,
        artist: song.artists[0].name,
        album: song.album.name,
        image: MetasImage.network(song.album.picUrl), //c
      );
    }
  }

  Future<void> seekBy(Duration by) async {
    await assetsAudioPlayer.seekBy(by);
  }

  Future<void> add(song) async {
    var audio = Audio.network(
      song.url,
      metas: Metas(
        title: song.name,
        artist: song.artists[0].name,
        album: song.album.name,
        image:
            MetasImage.network(song.album.picUrl), //can be MetasImage.network
      ),
    );
    assetsAudioPlayer.playlist.add(audio);
  }

  Future<void> reMoveAtIndex(int index) async {
    print(assetsAudioPlayer.playlist.audios.length);
    assetsAudioPlayer.playlist.removeAtIndex(index);
  }

  Future<void> playlistPlayAtIndex(int index) async {
    await assetsAudioPlayer.playlistPlayAtIndex(index);
  }

  Future<void> seek(Duration by) async {
    await assetsAudioPlayer.seek(by);
  }

  Future<void> playOrPause() async {
    await assetsAudioPlayer.playOrPause();
  }

  Future<void> play() async {
    await assetsAudioPlayer.play();
  }

  Future<void> pause() async {
    await assetsAudioPlayer.pause();
  }

  Future<void> stop() async {
    await assetsAudioPlayer.stop();
  }

  Future<void> next() async {
    await assetsAudioPlayer.next();
  }

  Future<void> prev() async {
    await assetsAudioPlayer.previous();
  }

  Future<void> dispose() async {
    await assetsAudioPlayer.dispose();
  }
}
```
Can you try again?

**Comment by johnny-stevie** [comment]
@liuchuancong Unfortunately, the same behaviour happens with your code. I don't actually see much difference in it also, tbh. You are just declaring the assetsAudioPlayer in a different place with an id that's all. Thanks for the suggestion anyway :)

**Comment by pparadox11** [comment]
there is definitely a problem, also if my app is on background and i am starting the player doesnt play the sound, 
as in version 1.6.3 i could do this on iOS

**Comment by johnny-stevie** [comment]
Anybody got any suggestions?

**Comment by spideythewebhead** [comment]
Personally I went back to 1.6.3 , and because I needed to play sound from the network, and the plugin doesn't have that ability in this version , I just manually download the file (not a suggestion, just maybe you could do something relative for you)

**Comment by johnny-stevie** [comment]
Unfortunately, I need to play network straight from source instead of
downloading it then playing it. I've thought about copying the kotlin code
from the package itself and tweak things that are wrong for it, but it's
the last thing I want to do since this package is extremely portable to IOS
and Web. If this issue doesn't resolve then I'll have to go with that
option.

**Comment by ashishfeels** [comment]
@johnny-stevie same behaviour when i play live stream also tried stopping the current player before staring new but if you tap very quickly to second stream it creates new player and player parallely

**Comment by florent37** [comment]
I try to reproduce, it only happen on network files when you call multiple times the open method ?

**Comment by florent37** [comment]
I think I miss a "semaphore" / "mutex" protection to avoid call open method before the previous has finished to

**Comment by ashishfeels** [comment]
> I try to reproduce, it only happen on network files when you call multiple times the open method ?
> in a short amount of time

yes,  i just copied what was there in the live stream documentations
@florent37 am i doing wrong ?
what should be the correct usage ?

**Comment by florent37** [comment]
nope, it's my fault, I do not have a protection of calling multiple time the open method

**Comment by florent37** [comment]
I'm working on it now to fix it

**Comment by florent37** [comment]
for my sample "livestream.dart" cannot allow multiple open (because it's called on init)

so i did not found it, thanks btw for the issue, it's with you I can make this package better & better <3

**Comment by ashishfeels** [comment]
@florent37 This package is awesome man, We should be thanking you.

**Comment by ashishfeels** [comment]
also is it possible to record or download the current stream to the device? i see there is some cache and download commits but wonder how it works.
@florent37

**Comment by florent37** [comment]
mmm you can fetch a song, then play it using Audio.network(url, cached: true), but for a livestream (with no end) I don't think it will work

**Comment by florent37** [comment]
please try the `2.0.8` I fixed the open method

**Comment by ashishfeels** [comment]
probably possible i have seen some radio app doing that but don't know how?

**Comment by ashishfeels** [comment]
> please try the `2.0.8` I fixed the open method

you are awesome

**Comment by johnny-stevie** [comment]
@florent37 This is fantastic! Thank you. The open function works completely. However the next() and previous() functions on network songs are behaving in the same way as .open() did which I don't understand. Looking at the package code, you still use the _openPlaylistCurrent function to achieve the same result. Btw, this happens when you spam the function as it did with .open(). I tried using .playlistPlayAtIndex() but same problem occurs, same with Future.delay() before calling .next()/.previous(). Maybe there's a quick fix that I missed so you don't have to reopen the issue again?

**Comment by ashishfeels** [comment]
@florent37 @johnny-stevie  Just checked it showing the similar behaviour on Livestream too tell me if you find anything.

**Comment by johnny-stevie** [comment]
@florent37 Any progress? Should I create a new issue for this? In my opinion I think that the next() and prev() are still linked to the open method problem which is why it should stay on this issue.

**Comment by safder2000** [comment]
I used  a 100 millisec delay before calling the function

  static playNextSong() async {
      await Future.delayed(const Duration(milliseconds: 100));
      try {
        await assetsAudioPlayer.next();
        await assetsAudioPlayer.play();
      } on Exception {
        log('Unsupported format');
      }
    }


  same for every other function ..
works for me.
